src/sort.rs

```diff
@@ -262,19 +340,97 @@ fn sort_by_ordering(
 ) {
     let mut idx = 0;
     for heading in ordering {
-        if let Some((_, to_sort_headings)) =
-            heading_order.iter().find(|((_, key), _)| key == heading)
-        {
-            for h in to_sort_headings {
-                if let Heading::Complete(segs) = h {
-                    let mut table = Some(toml.as_table_mut());
-                    for seg in segs {
-                        table = table.and_then(|t| t[seg].as_table_mut());
+        let mut matches: Vec<(&(usize, String), &Vec<Heading>)> = heading_order
+            .iter()
+            .filter(|((_s, key), headings)| {
+                key == heading
+                    || headings.iter().any(|h| {
+                        if let Heading::Complete(segs) = h {
+                            segs.iter().any(|seg| seg == heading)
+                        } else {
+                            false
+                        }
+                    })
+            })
+            .collect();
+
+        matches.sort_by(|a: &(&(usize, String), &Vec<Heading>), b| {
+            let a1_longest =
+                a.1.iter()
+                    .filter_map(|h| {
+                        if let Heading::Complete(segs) = h {
+                            if segs.iter().last() != Some(heading) {
+                                return None;
+                            }
+                            Some(segs.iter().rev().cloned().collect::<Vec<_>>().join("."))
+                        } else {
+                            None
+                        }
+                    })
+                    .max_by_key(|s| s.len());
+            let b1_longest =
+                b.1.iter()
+                    .filter_map(|h| {
+                        if let Heading::Complete(segs) = h {
+                            if segs.iter().last() != Some(heading) {
+                                return None;
+                            }
+                            Some(segs.iter().rev().cloned().collect::<Vec<_>>().join("."))
+                        } else {
+                            None
+                        }
+                    })
+                    .max_by_key(|s| s.len());
+
```

Comment:
Can we extract helper functions for this to de-duplicate the code?

Reply:
done